### ISSUE TITLE
Warnings as errors.

### ISSUE THREAD

**Issue by jbrezmorf** [flow123d/flow123d#1185]
- resolve current warnings in master
- set warnings as errors, e.g. see compiler parameters:
  -pedantic-errors (strict ISO C and ISO C++)
  -Wall (all warnings, can be relaxed by unsetting some special warnings)
  -Wextra (some extra warnings, see [discussion](https://stackoverflow.com/questions/2888404/wextra-how-useful-is-it-really))
  -Werror (warnings as errors)
- use this setting as default build parameters and on Jenkins
- turn off time critical test in `remove_mhdh`
- merge `remove_mhdh` into master

**Comment by Paulie14** [comment]
I do not understand this warning:
```
In file included from /home/paulie/docker_flow123d/flow123d/src/system/sys_profiler.cc:30:0:
/home/paulie/docker_flow123d/flow123d/src/system/sys_profiler.hh:218:50: warning: ‘__func__’ is not defined outside of function scope
 #define CODE_POINT(tag) CodePoint(tag, __FILE__, __func__, __LINE__)
                                                  ^
/home/paulie/docker_flow123d/flow123d/src/system/sys_profiler.cc:292:39: note: in expansion of macro ‘CODE_POINT’
 static CONSTEXPR_ CodePoint main_cp = CODE_POINT("Whole Program");
```

**Comment by jbrezmorf** [comment]
__func__ is an internal gcc macro that expands to the name of the function where it is expanded.
Bug introduced in : bff5e7340
@janhybs: I also wonder why we can not use the local static variable in the instance method to implement the Profiler singleton. I prefer this kind of encapsulation instead of the static class member.

**Comment by jbrezmorf** [comment]
Try to remove Profiler::initialize() completely and move the static variable with the Profiler instace 
into the Profiler::instance() method. As it was done before bff5e73.

**Comment by Paulie14** [comment]
- I resolved the profiler warning.

- I resolved the the "unused xxx" warnings. For unused function parameters I found, that we already have `FMT_UNUSED` macro, so I tried to use, where it made sense (inheritance - not all descendants use all the function parameters, templates - the same in specialization).
This really shows us some of "bad concept" and "forgotten" pieces of code.
I also made 2 issues #1197 and #1198 to consider class interface improvements, but more might be considered based on the resolved warnings..

- We should improve virtual-override keywords usage to improve code readability (I got lost in transport classes..). Warning `-Wsuggest-override` suggests adding `override` keyword, but it writes a looong list of warnings from other libs, e.g. mpi.

Using `pedantic` I got:
- `ISO C++ forbids zero-size array` in RefElement on the various "maps" arrays.
and
- `ISO C++ forbids variable length array [-Wvla]` in various places.
- `flow123d/src/system/logger.cc:40:42: warning: binary constants are a C++14 feature or GCC extension
StreamMask StreamMask::cerr = StreamMask(0b00000010);`
I am not sure, whether we have to deal with all these..

**Comment by jbrezmorf** [comment]
Great work. 

virtual vs. override - historically we were misusing `virtual` to mark override. suggest-override can be used temporarily to add some override keywords. Not critical.

I propose to use pedantic and turn it off only for the first types of error (using -Wno-error=...), 
possibly fixing the RefElement (?later). The third issue is gone as we are now moving to use C++14 standard.

**Comment by Paulie14** [comment]
- I switched several arrays to vectors in RefElement and cleaned up the rest of the warnings.
- the only left are the ones due to C++14.
- I updated the compiler warning flags in CMakeList (not sure, if it is necessary to do something else regarding Jenkins), once switched to C++14 standard, we can add `-Werror` to compiler flags.
- I have to increase the time limit for tests (if all pass, can be merged)

**Comment by Paulie14** [comment]
Working on merge to `remove_mhdh` branch, I get this warning on sized deallocation in profiler:

**Comment by Paulie14** [comment]
- merge master -> remove_mhdh done
- all warnings resolved in remove_mhdh
- Werror set in remove_mhdh

**Comment by Paulie14** [comment]
Error after compiling on Jenkins due to gtest:
```
[100%] Linking CXX static library ../lib/libflow123d_lib.a
make[4]: Leaving directory '/opt/flow123d/flow123d/build-remove_mhdh'
[100%] Built target flow123d_lib
make[4]: Entering directory '/opt/flow123d/flow123d/build-remove_mhdh'
Scanning dependencies of target flow123d
make[4]: Leaving directory '/opt/flow123d/flow123d/build-remove_mhdh'
make[4]: Entering directory '/opt/flow123d/flow123d/build-remove_mhdh'
[100%] Building CXX object CMakeFiles/bin/flow123d.dir/src/main.cc.o
[100%] Linking CXX executable bin/flow123d
make[4]: Leaving directory '/opt/flow123d/flow123d/build-remove_mhdh'
[100%] Built target bin/flow123d
make[3]: Leaving directory '/opt/flow123d/flow123d/build-remove_mhdh'
make[2]: Leaving directory '/opt/flow123d/flow123d/build-remove_mhdh'
make[1]: Leaving directory '/opt/flow123d/flow123d/build-remove_mhdh'
make: Leaving directory '/opt/flow123d/flow123d'
+ docker exec contdebug make -C /opt/flow123d/flow123d/build_tree/unit_tests -j4 gtest_mpi_obj
make: Entering directory '/opt/flow123d/flow123d/build-remove_mhdh/unit_tests'
make[1]: Entering directory '/opt/flow123d/flow123d/build-remove_mhdh'
make[2]: Entering directory '/opt/flow123d/flow123d/build-remove_mhdh'
make[3]: Entering directory '/opt/flow123d/flow123d/build-remove_mhdh'
Scanning dependencies of target gtest
make[3]: Leaving directory '/opt/flow123d/flow123d/build-remove_mhdh'
make[3]: Entering directory '/opt/flow123d/flow123d/build-remove_mhdh'
[  0%] Building CXX object third_party/gtest-1.7.0/CMakeFiles/gtest.dir/src/gtest-all.cc.o
[ 50%] Linking CXX static library libgtest.a
make[3]: Leaving directory '/opt/flow123d/flow123d/build-remove_mhdh'
[ 50%] Built target gtest
make[3]: Entering directory '/opt/flow123d/flow123d/build-remove_mhdh'
Scanning dependencies of target gtest_main
make[3]: Leaving directory '/opt/flow123d/flow123d/build-remove_mhdh'
make[3]: Entering directory '/opt/flow123d/flow123d/build-remove_mhdh'
[ 50%] Building CXX object third_party/gtest-1.7.0/CMakeFiles/gtest_main.dir/src/gtest_main.cc.o
[ 50%] Linking CXX static library libgtest_main.a
make[3]: Leaving directory '/opt/flow123d/flow123d/build-remove_mhdh'
[ 50%] Built target gtest_main
make[3]: Entering directory '/opt/flow123d/flow123d/build-remove_mhdh'
Scanning dependencies of target gtest_mpi_obj
make[3]: Leaving directory '/opt/flow123d/flow123d/build-remove_mhdh'
make[3]: Entering directory '/opt/flow123d/flow123d/build-remove_mhdh'
[100%] Building CXX object unit_tests/CMakeFiles/gtest_mpi_obj.dir/_gtest_extensions/gtest_mpi.cc.o
/opt/flow123d/flow123d/unit_tests/_gtest_extensions/gtest_mpi.cc:108:76: error: unused parameter ‘unit_test’ [-Werror=unused-parameter]
 void MPI_PrettyUnitTestResultPrinter::OnTestIterationStart(const UnitTest& unit_test, int iteration) {
                                                                            ^
/opt/flow123d/flow123d/unit_tests/_gtest_extensions/gtest_mpi.cc:108:91: error: unused parameter ‘iteration’ [-Werror=unused-parameter]
 void MPI_PrettyUnitTestResultPrinter::OnTestIterationStart(const UnitTest& unit_test, int iteration) {
                                                                                           ^
cc1plus: all warnings being treated as errors
unit_tests/CMakeFiles/gtest_mpi_obj.dir/build.make:62: recipe for target 'unit_tests/CMakeFiles/gtest_mpi_obj.dir/_gtest_extensions/gtest_mpi.cc.o' failed
make[3]: Leaving directory '/opt/flow123d/flow123d/build-remove_mhdh'
make[3]: *** [unit_tests/CMakeFiles/gtest_mpi_obj.dir/_gtest_extensions/gtest_mpi.cc.o] Error 1
CMakeFiles/Makefile2:1892: recipe for target 'unit_tests/CMakeFiles/gtest_mpi_obj.dir/all' failed
make[2]: Leaving directory '/opt/flow123d/flow123d/build-remove_mhdh'
CMakeFiles/Makefile2:1904: recipe for target 'unit_tests/CMakeFiles/gtest_mpi_obj.dir/rule' failed
make[2]: *** [unit_tests/CMakeFiles/gtest_mpi_obj.dir/all] Error 2
make[1]: *** [unit_tests/CMakeFiles/gtest_mpi_obj.dir/rule] Error 2
make[1]: Leaving directory '/opt/flow123d/flow123d/build-remove_mhdh'
Makefile:226: recipe for target 'unit_tests/CMakeFiles/gtest_mpi_obj.dir/rule' failed
make: Leaving directory '/opt/flow123d/flow123d/build-remove_mhdh/unit_tests'
make: *** [unit_tests/CMakeFiles/gtest_mpi_obj.dir/rule] Error 2
Build step 'Execute shell' marked build as failure
Set build name.
New build name is 'remove_mhdh@840821  #546'
Finished: FAILURE
```

- it builds something for unit test right after building Flow123d
- I don't know how to control which flags are used where (especially for unit tests, which full of warnings - especially from gtest)
- I will switch off the Werror for now

**Comment by jbrezmorf** [comment]
'gtest_mpi.cc' is our own extension so we can fix it

**Comment by jbrezmorf** [comment]
I also did some changes related to the gtest in master. I have upgraded gtest_1.7 to gtest 1.10 and do some related fixes.

**Comment by Paulie14** [comment]
- Werror set by 1df992a38a97487b7fd7c6a412fca90228ffd219 in `remove_mh_dh`
- long test 22-4 fixed by b2051dd42284467fb9c48d2100fd15e2e61a7a3d

I suppose that everything ready to merge `remove_mhdh` -> `master`

**Comment by Paulie14** [comment]
Note: the only unit test failing is still in system Profiler memory test

**Comment by Paulie14** [comment]
system Profiler memory unit test temporarily "greened"  by56a6dc3ee3c3f7db8bc967a27e09df85292278e6

everything ready to merge `remove_mhdh` -> `master`

**Comment by Paulie14** [comment]
Merged c471082b4839c336a7b8360726f007da7e3a5895.
